client/gui/InterfaceObjectConfigurable.cpp

```diff
@@ -0,0 +1,386 @@
+	auto style = config["style"].String() == "brown" ? CSlider::BROWN : CSlider::BLUE;
+	auto itemsVisible = config["itemsVisible"].Integer();
+	auto itemsTotal = config["itemsTotal"].Integer();
+	auto value = config["selected"].Integer();
+	bool horizontal = config["orientation"].String() == "horizontal";
+	return std::make_shared<CSlider>(position, length, callbacks.at(config["callback"].String()), itemsVisible, itemsTotal, value, horizontal, style);
+}
+
+std::shared_ptr<CAnimImage> InterfaceObjectConfigurable::buildImage(const JsonNode & config) const
+{
+	auto position = readPosition(config["position"]);
+	auto image = config["image"].String();
+	int group = config["group"].isNull() ? 0 : config["group"].Integer();
+	int frame = config["frame"].isNull() ? 0 : config["frame"].Integer();
+	return std::make_shared<CAnimImage>(image, frame, group, position.x, position.y);
+}
+
+std::shared_ptr<CFilledTexture> InterfaceObjectConfigurable::buildTexture(const JsonNode & config) const
+{
+	auto image = config["image"].String();
+	auto rect = readRect(config["rect"]);
+	return std::make_shared<CFilledTexture>(image, rect);
+}
+
+std::shared_ptr<CShowableAnim> InterfaceObjectConfigurable::buildAnimation(const JsonNode & config) const
+{
+	auto position = readPosition(config["position"]);
+	auto image = config["image"].String();
+	ui8 flags = 0;
+	if(!config["repeat"].Bool())
+		flags |= CShowableAnim::EFlags::PLAY_ONCE;
+	
+	int group = config["group"].isNull() ? 0 : config["group"].Integer();
+	auto anim = std::make_shared<CShowableAnim>(position.x, position.y, image, flags, 4, group);
+	if(!config["alpha"].isNull())
+		anim->setAlpha(config["alpha"].Integer());
+	if(!config["callback"].isNull())
+		anim->callback = std::bind(callbacks.at(config["callback"].String()), 0);
+	if(!config["frames"].isNull())
+	{
+		auto b = config["frames"]["start"].Integer();
+		auto e = config["frames"]["end"].Integer();
+		anim->set(group, b, e);
+	}
+	return anim;
+}
+
+std::shared_ptr<CIntObject> InterfaceObjectConfigurable::buildWidget(JsonNode config) const
+{
+	assert(!config.isNull());
+	//overrides from variables
+	for(auto & item : config["overrides"].Struct())
+	{
+		config[item.first] = variables[item.second.String()];
+	}
+	
+	auto type = config["type"].String();
+	if(type == "picture")
+	{
+		return buildPicture(config);
+	}
+	if(type == "image")
+	{
+		return buildImage(config);
+	}
+	if(type == "texture")
+	{
+		return buildTexture(config);
+	}
+	if(type == "animation")
+	{
+		return buildAnimation(config);
+	}
+	if(type == "label")
+	{
+		return buildLabel(config);
+	}
+	if(type == "toggleGroup")
+	{
+		return buildToggleGroup(config);
+	}
+	if(type == "toggleButton")
+	{
+		return buildToggleButton(config);
+	}
+	if(type == "button")
+	{
+		return buildButton(config);
+	}
+	if(type == "labelGroup")
+	{
+		return buildLabelGroup(config);
+	}
+	if(type == "slider")
+	{
+		return buildSlider(config);
+	}
+	if(type == "custom")
```

Comment:
Perhaps instead of using "custom" type `buildCustomWidget` should be called unconditionally? (in cases when type was not caught by this if-forest)
This way, `buildCustomWidget` will be able to define its own widgets with any name, with less awkward syntax in json.  Useful if UI needs several different custom widgets.

Reply:
but it's type, not name. Name is separate field, so you can have many custom widgets with their specific names. And making them unconditionally can bring compatibility problems in future (when some version supports some new widgets)

Reply:
Sorry, not "name" but "type".
If you're creating items for dropbox it makes sense for these items to have "dropboxItem" type (instead of "custom").

And some classes (adventure map or battles) you may end up with several custom widgets, e.g. battle queue, battle log.

So it makes sense to allow derived classes to inject their own widgets with separate "type", e.g. `"type" : "battleLog"` even though these types are not defined in basic interface builder class.

Reply:
There is problem with that - when modders will see into those examples they may think that it's possible to create "dropboxitem" in their new configs. But it won't as processing "custom" objects is implemented in override function

Reply:
Hmm... could be solved by using better names, e.g. "rmgTemplateItem", indicating that this item was designed specifically for rmg templates.
"custom" is even worse using the same logic - modders may try to use parameters from one "custom" element in another.

It also very... nondescript. What exactly is "custom"? How it appears in UI? How can it be configured? Better name will provide hints by itself.

Reply:
But how it limits you? You still have all fields from object. You can define and process as many objects inside custom processing as you want. Custom means that object is not a standard but specific. As for me it doesn't confuse.

Reply:
Personally I prefer concrete/strongly defined terms/forms, Meanwhile, "custom" is very generic. Sure, it is flexible, but not sure if there is need for such flexibility.

Anyway, don't see any point in arguing. This is not something that will be set in stone, for now let's keep it as it.
I'm sure that some design issues will arise later anyway once this UI customization has been be expanded to other UI's.

Reply:
I took a look at config json and now agree with you - having multiple "custom" objects without any identification is misleading. Fixed